main/GUB2.c

```diff
@@ -0,0 +1,309 @@
+    // Forces the isr_core_id in the gpio_context to core 1. This is a hack.
+    gpio_intr_enable(PIN_NUM_CAN1_RX_INT);
+    gpio_intr_disable(PIN_NUM_CAN1_RX_INT);
+}
+
+/**
+ * Setup the main GUB Control interface
+*/
+void GUBInit(){
+    gubState.gubEvents = xEventGroupCreate();
+
+    //setup CAN SPI bus
+    ESP_LOGD(TAG, "Setting CAN up SPI bus");
+    
+    spi_bus_config_t canBuscfg={
+        .miso_io_num=PIN_NUM_CAN_MISO,
+        .mosi_io_num=PIN_NUM_CAN_MOSI,
+        .sclk_io_num=PIN_NUM_CAN_CLK,
+        .quadwp_io_num=-1,
+        .quadhd_io_num=-1,
+        .max_transfer_sz=0
+    };
+
+    ESP_ERROR_CHECK(spi_bus_initialize(CAN_SPI_HOST, &canBuscfg, SPI_DMA_CH_AUTO));
+
+    // Install GPIO service on Core 1, all gpio isr handlers will be processed on core 1.
+    esp_ipc_call_blocking(1, installGPIOISRService, 0);
+    gpio_install_isr_service(0);
+    
+    ESP_LOGD(TAG, "Starting CAN.");
+    // CAN bus driver setup. Don't want to miss anything so do this first!
+    canDriverInit(gubState.gubEvents, CAN_EVENT);
+    canDriverAddBus(0, PIN_NUM_CAN1_CS, PIN_NUM_CAN1_RX_INT, PIN_NUM_CAN1_STB);
+    
+    ESP_LOGI(TAG, "Initializing SD Card SPI bus");
+    spi_bus_config_t sdBusCFG = {
+        .mosi_io_num = PIN_NUM_SD_MOSI,
+        .miso_io_num = PIN_NUM_SD_MISO,
+        .sclk_io_num = PIN_NUM_SD_CLK,
+        .quadwp_io_num = -1,
+        .quadhd_io_num = -1,
+        .max_transfer_sz = 4092,
+    };
+    
+    //initialize SPI bus
+    esp_err_t ret;
+    ret = spi_bus_initialize(SD_SPI_HOST, &sdBusCFG, SDSPI_DEFAULT_DMA);
+    if (ret == ESP_OK) {
+        int r = GUBMountSDCard(SD_CARD_BASE_PATH);
+        ESP_LOGI("SD_CARD", "SD card initialized with status %d", r);
+        if(!r){
+            sdmmc_card_print_info(stdout, gubState.SDcard);
+        }
+    } else {
+        ESP_LOGE(TAG, "Failed to initialize bus.");
+    }
+
+    listDir(SD_CARD_BASE_PATH);
+
+    //start up logger
+    canLoggerInit();
+
+    GUBInitLED();
+
+    ESP_LOGI(TAG, "GUB Setup, starting main loop");
+    xTaskCreate( GUBloop, "GUB", GUB_STACK_SIZE, NULL, 2, &gubState.mainTaskHandler);
+}
+
+/**
+ * Main loop for GUB task
+*/
+void GUBloop(void *pvParam){
+
+    while (true)
+    {   
+        GUBHeartbeatUpdate();
+        checkSDCardStatus();
+        if(gubState.sdCardState == SD_NOT_MOUNTED && esp_timer_get_time() - gubState.lastMountAttempt > 1000000){ // check every second
+            gubState.lastMountAttempt = esp_timer_get_time();
+            ESP_LOGW(TAG, "SD card not mounted! Attempting to remount...");
+            GUBMountSDCard(SD_CARD_BASE_PATH);
+        }
+
+        canDriverUpdate();
+        canLoggerUpdate();
+        
+        CANMessage_t message;
+        if(canReceiveMessage(&message, pdMS_TO_TICKS(5))){
+            canLoggerProcessMessage(&message);
+        }
+        
+        //* needed if not relying to the timeout of the queue
+        // vTaskDelay(pdMS_TO_TICKS(5));
```

Comment:
It does as it should. At the full load I was able to test with (~1800/s) I was pushing 34% usage. This is significantly better than the previous method I had tested with events. It should not be a problem since there is still a bit of a margin. The bigger problem would likely be spikes in latency cause by the SDCard.